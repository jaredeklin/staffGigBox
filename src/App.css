

.DayPicker-Month {
  width: 400px;
  height: 300px;
}

header {
  display: flex;
  justify-content: space-between;
  height: 100px;
  padding: 20px;
  background-color: navy;
  color: #FFF;
}

.user-img {
  width: 60px;
  border-radius: 50%;
}


