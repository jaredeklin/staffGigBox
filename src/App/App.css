.DayPicker-Month {
  width: 400px;
  height: 300px;
}

ul {
  list-style: none;
  padding: 0;
}

.app {
  display: grid;
  grid-template-columns: 220px auto;
}

aside {
  grid-column: 1;
  padding: 20px;
  position: fixed;
  width: 220px;
}

.main-container {
  grid-column: 2;
  background-color: #ffffff;
  padding: 30px;
  min-height: 100vh;
}
