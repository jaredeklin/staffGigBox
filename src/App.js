import React, { Component } from 'react';
import './App.css'

import DayPicker, { DateUtils } from 'react-day-picker';
import 'react-day-picker/lib/style.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedDays: [],
    };

    this.availability = []
  }

  handleDayClick = (day, { selected }) => {
    const { selectedDays } = this.state;

    if (selected) {
      const selectedIndex = selectedDays.findIndex(selectedDay =>
        DateUtils.isSameDay(selectedDay, day)
      );
      selectedDays.splice(selectedIndex, 1);
    } else {
      selectedDays.push(day);
    }
    this.setState({ selectedDays });
  }

  handleSubmit = () => {
    this.availability = [...this.state.selectedDays]
    console.log(this.availability)
    /// POST request adding availability to staff members
  }

  render() {
    return (
      <div>
        <h1>Staff Gig Box</h1>
        <section className='calendar'>
          <DayPicker
            selectedDays={this.state.selectedDays}
            onDayClick={this.handleDayClick}
            className='calendar1'
          />
        </section>
        <button onClick={this.handleSubmit}>Submit</button>
      </div>
    );
  }
}

export default App;
