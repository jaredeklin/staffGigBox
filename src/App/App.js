import React, { Component } from 'react';
import './App.css'
import { Header } from '../Header/Header'
import { TabContainer } from '../TabContainer/TabContainer'
import { Api } from '../Api';



class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user: null,
      staff: [],
      events: [],
      schedule: [],
      isCurrentStaff: false,
      addNewStaff: false
    };

    this.api = new Api()
  }

  addUser = async (user) => {
    const { staff } = this.state

    await this.setState({ user, isCurrentStaff: false })

    if (user) {
      const match = staff.find(person => person.google_id === user.uid)

      if( match) {
        this.setState({ isCurrentStaff: true })
      } else {
        this.setState({ addNewStaff: true })
      }
    }
  }

  deleteFromSchedule = async (id) => {

    await fetch(`http://localhost:3000/api/v1/schedule/${id}`, {
      method: 'DELETE'
    })

    const schedule = await this.api.getSchedule()

    this.setState({ schedule })
  }

  editSchedule = async () => {
    const schedule = await this.api.getSchedule()

    this.setState({ schedule })
  }

  addStaff = () => {
    this.setState({
      isCurrentStaff: true,
      addNewStaff: false
    })
  }

  scheduleGenerator = async () => {
    const { staff, events } = this.state
    const generatedSchedule = this.api.generateSchedule(staff, events)

    await this.api.postSchedule(generatedSchedule)
    const schedule = await this.api.getSchedule()

    this.setState({ schedule })
  }


  updateStateFromHelpers = async () => {
    const staff = await this.api.getStaff()
    const events = await this.api.getEvents()
    const schedule = await this.api.getSchedule()

    this.setState({ staff, events, schedule })
  }

  componentDidMount = () => {
    this.updateStateFromHelpers()
  }

  render() {

    return (
      <div className='app'>
        <Header addUser={ this.addUser }/>
        <TabContainer
          editSchedule = { this.editSchedule }
          schedule={ this.state.schedule }
          scheduleGenerator={ this.scheduleGenerator }
          staff={ this.state.staff }
          addStaff={ this.addStaff }
          user={ this.state.user }
          deleteFromSchedule= { this.deleteFromSchedule }
        />
      </div>
    );
  }
}

export default App;
