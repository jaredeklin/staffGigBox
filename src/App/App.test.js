import React from 'react';
import App from './App';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;
  let mockUser1;

  beforeEach(() => {
    wrapper = shallow(<App />, { disableLifecycleMethods: true });
    mockUser1 = {
      uid: 12345
    };
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('addUser', () => {
    it('should update state and call checkAuthorization when there is a user', async () => {
      const mockStaff = [
        {
          google_id: 12345
        }
      ];

      wrapper.instance().checkAuthorization = jest.fn();
      wrapper.setState({ staff: mockStaff });
      await wrapper.instance().addUser(mockUser1);

      expect(wrapper.state('user')).toEqual(mockUser1);
      expect(wrapper.state('isCurrentStaff')).toEqual(false);
      expect(wrapper.instance().checkAuthorization).toHaveBeenCalledWith({
        google_id: 12345
      });
    });

    it('should set admin and tabs state if not a user ', async () => {
      await wrapper.instance().addUser();

      expect(wrapper.state('tabs')).toEqual(['Schedule']);
      expect(wrapper.state('admin')).toEqual(false);
    });
  });

  describe('checkAuthorization', () => {
    it('should set state with correct values when authorized', () => {
      wrapper.instance().checkAuthorization({ google_id: 12345, id: 3 });

      expect(wrapper.state('isCurrentStaff')).toEqual(true);
      expect(wrapper.state('tabs')).toEqual([
        'Schedule',
        'Submit Availability'
      ]);
      expect(wrapper.state('admin')).toEqual(false);
      expect(wrapper.state('currentUserId')).toEqual(3);

      wrapper
        .instance()
        .checkAuthorization({ google_id: 12345, id: 3, bar_manager: true });

      expect(wrapper.state('isCurrentStaff')).toEqual(true);
      expect(wrapper.state('tabs')).toEqual([
        'Schedule',
        'Submit Availability',
        'Add Event',
        'Add New Staff'
      ]);
      expect(wrapper.state('admin')).toEqual(true);
      expect(wrapper.state('currentUserId')).toEqual(3);
    });

    it('should set state with correct values when not authorized', () => {
      wrapper.instance().checkAuthorization();

      expect(wrapper.state('addNewStaff')).toEqual(true);
      expect(wrapper.state('tabs')).toEqual(['Schedule', 'Add New Staff']);
    });
  });

  it('should add Staff', () => {
    wrapper.instance().addStaff();
    expect(wrapper.state('isCurrentStaff')).toEqual(true);
    expect(wrapper.state('addNewStaff')).toEqual(false);
  });

  it('should delete userId and event Id from schedule', async () => {
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        status: 200,
        json: () =>
          Promise.resolve({
            mockUser1
          })
      })
    );
    const expected = [
      'http://localhost:3000/api/v1/schedule/1',
      {
        method: 'DELETE'
      }
    ];

    wrapper.instance().api.getSchedule = jest.fn();

    await wrapper.instance().deleteFromSchedule(1);

    expect(window.fetch).toHaveBeenCalledWith(...expected);
  });

  it('should set the state with new schedule', async () => {
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        status: 200,
        json: () =>
          Promise.resolve({
            mockUser1
          })
      })
    );
    wrapper.instance().editSchedule = jest.fn();

    await wrapper.instance().deleteFromSchedule(1);

    expect(wrapper.instance().editSchedule).toHaveBeenCalled();
  });

  it('should edit the schedule', async () => {
    const mockSchedule = {
      event_id: 1,
      date: '05/16/18',
      name: 'Sparklehorse'
    };

    wrapper.instance().api.getSchedule = jest
      .fn()
      .mockReturnValue(mockSchedule);

    await wrapper.instance().editSchedule();

    expect(wrapper.state('schedule')).toEqual(mockSchedule);
  });

  it('should generate a schedule', async () => {
    wrapper.instance().api.generateSchedule = jest.fn().mockReturnValue(true);
    wrapper.instance().api.modifySchedule = jest.fn();
    wrapper.instance().editSchedule = jest.fn();

    await wrapper.instance().scheduleGenerator();

    expect(wrapper.instance().api.generateSchedule).toHaveBeenCalled();
    expect(wrapper.instance().api.modifySchedule).toHaveBeenCalled();
    expect(wrapper.instance().editSchedule).toHaveBeenCalled();
  });

  it('should update the state with helpers', async () => {
    const mockStaff = { name: 'taco' };
    const mockEvent = { event: 'idone' };
    const mockSchedule = { schedule: 'thebesten' };

    wrapper.instance().api.getStaff = jest.fn().mockReturnValue(mockStaff);
    wrapper.instance().api.getEvents = jest.fn().mockReturnValue(mockEvent);
    wrapper.instance().api.getSchedule = jest
      .fn()
      .mockReturnValue(mockSchedule);

    await wrapper.instance().updateStateFromHelpers();

    expect(wrapper.instance().api.getStaff).toHaveBeenCalled();
    expect(wrapper.instance().api.getEvents).toHaveBeenCalled();
    expect(wrapper.instance().api.getSchedule).toHaveBeenCalled();

    expect(wrapper.state()).toEqual({
      addNewStaff: false,
      admin: false,
      tabs: [],
      events: { event: 'idone' },
      isCurrentStaff: false,
      schedule: { schedule: 'thebesten' },
      staff: { name: 'taco' },
      user: null,
      currentUserId: null
    });
  });

  it('should call updateStateFromHelpers in componentDidMount', () => {
    wrapper.instance().updateStateFromHelpers = jest.fn();

    wrapper.instance().componentDidMount();

    expect(wrapper.instance().updateStateFromHelpers).toHaveBeenCalled();
  });
});
