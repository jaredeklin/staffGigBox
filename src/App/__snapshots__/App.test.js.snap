// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Sidebar
          addUser={[Function]}
/>,
        <div
          className="main-container"
>
          <withRouter(Header) />
          <div
                    className="main-display-container"
          >
                    <Routes
                              appState={
                                        Object {
                                                  "addNewStaff": false,
                                                  "admin": false,
                                                  "currentUser": Object {},
                                                  "events": Array [],
                                                  "isCurrentStaff": false,
                                                  "schedule": Array [],
                                                  "staff": Array [],
                                                  "unscheduledEvents": Array [],
                                                }
                              }
                              methods={
                                        Object {
                                                  "addEvent": [Function],
                                                  "addStaff": [Function],
                                                  "deleteFromSchedule": [Function],
                                                  "editSchedule": [Function],
                                                  "scheduleGenerator": [Function],
                                                }
                              }
                    />
          </div>
</div>,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "addUser": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <withRouter(Header) />,
            <div
              className="main-display-container"
>
              <Routes
                            appState={
                                          Object {
                                                        "addNewStaff": false,
                                                        "admin": false,
                                                        "currentUser": Object {},
                                                        "events": Array [],
                                                        "isCurrentStaff": false,
                                                        "schedule": Array [],
                                                        "staff": Array [],
                                                        "unscheduledEvents": Array [],
                                                      }
                            }
                            methods={
                                          Object {
                                                        "addEvent": [Function],
                                                        "addStaff": [Function],
                                                        "deleteFromSchedule": [Function],
                                                        "editSchedule": [Function],
                                                        "scheduleGenerator": [Function],
                                                      }
                            }
              />
</div>,
          ],
          "className": "main-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Routes
                appState={
                                Object {
                                                "addNewStaff": false,
                                                "admin": false,
                                                "currentUser": Object {},
                                                "events": Array [],
                                                "isCurrentStaff": false,
                                                "schedule": Array [],
                                                "staff": Array [],
                                                "unscheduledEvents": Array [],
                                              }
                }
                methods={
                                Object {
                                                "addEvent": [Function],
                                                "addStaff": [Function],
                                                "deleteFromSchedule": [Function],
                                                "editSchedule": [Function],
                                                "scheduleGenerator": [Function],
                                              }
                }
/>,
              "className": "main-display-container",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "appState": Object {
                  "addNewStaff": false,
                  "admin": false,
                  "currentUser": Object {},
                  "events": Array [],
                  "isCurrentStaff": false,
                  "schedule": Array [],
                  "staff": Array [],
                  "unscheduledEvents": Array [],
                },
                "methods": Object {
                  "addEvent": [Function],
                  "addStaff": [Function],
                  "deleteFromSchedule": [Function],
                  "editSchedule": [Function],
                  "scheduleGenerator": [Function],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Sidebar
            addUser={[Function]}
/>,
          <div
            className="main-container"
>
            <withRouter(Header) />
            <div
                        className="main-display-container"
            >
                        <Routes
                                    appState={
                                                Object {
                                                            "addNewStaff": false,
                                                            "admin": false,
                                                            "currentUser": Object {},
                                                            "events": Array [],
                                                            "isCurrentStaff": false,
                                                            "schedule": Array [],
                                                            "staff": Array [],
                                                            "unscheduledEvents": Array [],
                                                          }
                                    }
                                    methods={
                                                Object {
                                                            "addEvent": [Function],
                                                            "addStaff": [Function],
                                                            "deleteFromSchedule": [Function],
                                                            "editSchedule": [Function],
                                                            "scheduleGenerator": [Function],
                                                          }
                                    }
                        />
            </div>
</div>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "addUser": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <withRouter(Header) />,
              <div
                className="main-display-container"
>
                <Routes
                                appState={
                                                Object {
                                                                "addNewStaff": false,
                                                                "admin": false,
                                                                "currentUser": Object {},
                                                                "events": Array [],
                                                                "isCurrentStaff": false,
                                                                "schedule": Array [],
                                                                "staff": Array [],
                                                                "unscheduledEvents": Array [],
                                                              }
                                }
                                methods={
                                                Object {
                                                                "addEvent": [Function],
                                                                "addStaff": [Function],
                                                                "deleteFromSchedule": [Function],
                                                                "editSchedule": [Function],
                                                                "scheduleGenerator": [Function],
                                                              }
                                }
                />
</div>,
            ],
            "className": "main-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Routes
                  appState={
                                    Object {
                                                      "addNewStaff": false,
                                                      "admin": false,
                                                      "currentUser": Object {},
                                                      "events": Array [],
                                                      "isCurrentStaff": false,
                                                      "schedule": Array [],
                                                      "staff": Array [],
                                                      "unscheduledEvents": Array [],
                                                    }
                  }
                  methods={
                                    Object {
                                                      "addEvent": [Function],
                                                      "addStaff": [Function],
                                                      "deleteFromSchedule": [Function],
                                                      "editSchedule": [Function],
                                                      "scheduleGenerator": [Function],
                                                    }
                  }
/>,
                "className": "main-display-container",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "appState": Object {
                    "addNewStaff": false,
                    "admin": false,
                    "currentUser": Object {},
                    "events": Array [],
                    "isCurrentStaff": false,
                    "schedule": Array [],
                    "staff": Array [],
                    "unscheduledEvents": Array [],
                  },
                  "methods": Object {
                    "addEvent": [Function],
                    "addStaff": [Function],
                    "deleteFromSchedule": [Function],
                    "editSchedule": [Function],
                    "scheduleGenerator": [Function],
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
